### PR TITLE
IGNITE-16269 Added max rows support.

### PR DESCRIPTION
**Pull request by vladErmakov07**

### CODE REVIEW

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/sql/JdbcQueryCursor.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    @Override
+    public T next() {
+        int fetchSize = (maxRows > 0) ? (int) Math.min(1, maxRows - fetched) : 1;
+
+        if (fetchSize > 0) {
+            fetched++;
+
+            return iter.next();
+        }
+        return null;
```
Comment (AMashenkov) [inline modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/sql/JdbcQueryCursor.java:80]: Should the cursor be closed immediately?

**Review by AMashenkov — APPROVED** [review]

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/JdbcQueryEventHandlerImpl.java [hunk 2]
```diff
@@ -93,7 +94,12 @@ public JdbcQueryEventHandlerImpl(QueryProcessor processor, JdbcMetadataCatalog m
 
         List<SqlCursor<List<?>>> cursors;
         try {
-            cursors = processor.query(req.schemaName(), req.sqlQuery(), req.arguments() == null ? new Object[0] : req.arguments());
+            List<SqlCursor<List<?>>> queryCursors = processor.query(req.schemaName(), req.sqlQuery(),
+                    req.arguments() == null ? new Object[0] : req.arguments());
```
Comment (korlov42) [inline modules/client-handler/src/main/java/org/apache/ignite/client/handler/JdbcQueryEventHandlerImpl.java:98]: org.apache.ignite.internal.util.ArrayUtils#OBJECT_EMPTY_ARRAY

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/sql/JdbcQueryCursor.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+ */
+public class JdbcQueryCursor<T> implements SqlCursor<T> {
+    /** Max rows. */
+    private final long maxRows;
+
+    /** Query result rows. */
+    private final SqlCursor<T> cur;
+
+    /** Query results iterator. */
+    private final Iterator<T> iter;
```
Comment (korlov42) [inline modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/sql/JdbcQueryCursor.java:36]: why do you need an iterator? SqlCursor implements Iterator, so you could use it directly

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/sql/JdbcQueryCursor.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    public T next() {
+        int fetchSize = (maxRows > 0) ? (int) Math.min(1, maxRows - fetched) : 1;
+
+        if (fetchSize > 0) {
+            fetched++;
+
+            return iter.next();
+        }
+        return null;
+    }
```
Comment (korlov42) [inline modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/sql/JdbcQueryCursor.java:81]:
```suggestion
    public T next() {
        if (!hasNext()) {
            throw new NoSuchElementException();
        }

        T res = cur.next();

        fetched++;

        return res;
    }
```

please note that throwing an exception is required by iterator's contract. See java.util.Iterator#next

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/sql/JdbcQueryCursor.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    /** Query results iterator. */
+    private final Iterator<T> iter;
+
+    /** Number of fetched rows. */
+    private long fetched;
+
+    /**
+     * Constructor.
+     *
+     * @param maxRows Max rows.
```
Comment (korlov42) [inline modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/sql/JdbcQueryCursor.java:44]: looks like iterator's behaviour is dependent on the particular value passed as the `maxRows` param. Could you please describe the difference in the Javadoc? Does it make any sense to prohibit any negative values?

Comment (vladErmakov07) [reply modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/sql/JdbcQueryCursor.java:48]: I have added additional javadocs detailing how the class works. 
I don't think we should ban negative values. The user sets the maxRows value with Statement#setMaxRows. According to the contract of this class, the function throws an exception if the value is less than zero.

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/sql/JdbcQueryCursor.java (hunk 1 above)
Comment (vladErmakov07) [reply modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/sql/JdbcQueryCursor.java]: This situation is not possible after the fixes. The method now throws an exception if the return value does not exist to match the method contract.

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/sql/JdbcQueryCursor.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import org.apache.ignite.internal.sql.engine.ResultSetMetadata;
+import org.apache.ignite.internal.sql.engine.SqlCursor;
+import org.apache.ignite.internal.sql.engine.SqlQueryType;
+
+/**
+ * Sql query cursor with the ability to limit the maximum number of rows returned.
+ * The maxRows parameter is responsible for the number of rows returned.
```
Comment (AMashenkov) [inline modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/sql/JdbcQueryCursor.java:28]:
```suggestion
 * The  {@link JdbcQueryCursor#maxRows} parameter is responsible for the number of rows returned.
```

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/sql/JdbcQueryCursor.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    public boolean hasNext() {
+        return cur.hasNext() && !(maxRows > 0 && fetched >= maxRows);
+    }
+
+    /**
+     * Returns the next element in the iteration if the iteration has more elements
+     * and the limit of maxRows has not been reached.
+     *
+     * @return the next element if the cursor has more rows and the limit of the maximum rows hasn't been reached.
+     */
```
Comment (AMashenkov) [inline modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/sql/JdbcQueryCursor.java:73]:
```suggestion
     * @throws NoSuchElementException if the iteration has no more elements.
     */
```

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/sql/JdbcQueryCursor.java [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+        T res = cur.next();
+
+        fetched++;
+
+        return res;
+    }
+
+    /** {@inheritDoc} */
+    @Override public void close() throws Exception {
```
Comment (AMashenkov) [inline modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/sql/JdbcQueryCursor.java:88]:
```suggestion
    @Override 
    public void close() throws Exception {
```

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/sql/JdbcQueryCursor.java [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+import org.apache.ignite.internal.sql.engine.SqlQueryType;
+
+/**
+ * Sql query cursor with the ability to limit the maximum number of rows returned.
+ * The maxRows parameter is responsible for the number of rows returned.
+ * It can either be equal to zero or have a positive value.
+ * Zero means that there are no additional restrictions other than the hasNext function,
+ * and the cursor behaves like a normal iterator.
+ * A positive value means that the cursor will return values either until the hasNext function
+ * returns false, or until the number of records already returned equals maxRows.
```
Comment (AMashenkov) [inline modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/sql/JdbcQueryCursor.java:34]:
```suggestion
 *
 * <p>The  {@link JdbcQueryCursor#maxRows} parameter limits the amount of rows to be returned by the cursor.
 * It value can either be a positive value or equal to zero, where zero means no limit.
```

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/sql/JdbcQueryCursor.java [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+    /** Query result rows. */
+    private final SqlCursor<T> cur;
+
+    /** Number of fetched rows. */
+    private long fetched;
+
+    /**
+     * Constructor.
+     *
+     * @param maxRows Max rows.
```
Comment (AMashenkov) [inline modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/sql/JdbcQueryCursor.java:49]:
```suggestion
     * @param maxRows Max amount of rows cursor will return, or zero if unlimited.
```

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/sql/JdbcQueryCursor.java [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import org.apache.ignite.internal.sql.engine.ResultSetMetadata;
+import org.apache.ignite.internal.sql.engine.SqlCursor;
+import org.apache.ignite.internal.sql.engine.SqlQueryType;
+
+/**
+ * Sql query cursor with the ability to limit the maximum number of rows returned.
+ *
+ *
```
Comment (AMashenkov) [inline modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/sql/JdbcQueryCursor.java:29]:
```suggestion
 *
```

**Pull request merged**